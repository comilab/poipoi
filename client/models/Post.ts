import dayjs, { Dayjs } from 'dayjs'

import Image from '~/models/Image'
import Reaction from '~/models/Reaction'

export type ScopeUnion = 'public'|'password'|'private'

export type RatingUnion = null|'nsfw'|'r18'

export default class Post {
  id: number
  userId: number
  title: string
  caption: string
  showImagesList: boolean
  text: string
  scope: ScopeUnion
  password: string|null
  publishStart: Dayjs|null = null
  publishEnd: Dayjs|null = null
  rating: RatingUnion
  pinned: boolean
  showThumbnail: boolean
  denyRobot: boolean|null
  enableReaction: boolean|null
  allowedEmojis: string[]|null
  deniedEmojis: string[]|null
  enableTwitterShare: boolean|null
  createdAt: Dayjs|null = null
  updatedAt: Dayjs|null = null

  url: string
  textCount: number
  actualDenyRobot: boolean
  actualEnableReaction: boolean
  actualAllowedEmojis: string[]
  actualDeniedEmojis: string[]
  actualEnableTwitterShare: boolean
  emojis: string[] = []

  images: Image[] = []
  imagesCount: number

  reactions: Reaction[] = []
  reactionsCount: number

  constructor (data: any) {
    this.id = data.id
    this.userId = data.userId
    this.title = data.title
    this.caption = data.caption
    this.showImagesList = data.showImagesList
    this.text = data.text
    this.scope = data.scope
    this.password = data.password
    if (data.publishStart) {
      this.publishStart = dayjs.utc(data.publishStart).local()
    }
    if (data.publishEnd) {
      this.publishEnd = dayjs.utc(data.publishEnd).local()
    }
    this.rating = data.rating
    this.pinned = data.pinned
    this.showThumbnail = data.showThumbnail
    this.denyRobot = data.denyRobot
    this.enableReaction = data.enableReaction
    this.allowedEmojis = data.allowedEmojis
    this.deniedEmojis = data.deniedEmojis
    this.enableTwitterShare = data.enableTwitterShare
    if (data.createdAt) {
      this.createdAt = dayjs.utc(data.createdAt).local()
    }
    if (data.updatedAt) {
      this.updatedAt = dayjs.utc(data.updatedAt).local()
    }

    this.url = data.url
    this.textCount = data.textCount
    this.actualDenyRobot = data.actualDenyRobot
    this.actualEnableReaction = data.actualEnableReaction
    this.actualAllowedEmojis = data.actualAllowedEmojis
    this.actualDeniedEmojis = data.actualDeniedEmojis
    this.actualEnableTwitterShare = data.actualEnableTwitterShare
    this.emojis = data.emojis

    if (data.images) {
      this.images = data.images.map((image: any) => new Image(image))
    }
    this.imagesCount = data.imagesCount

    if (data.reactions) {
      this.reactions = data.reactions.map((reaction: any) => new Reaction(reaction))
    }
    this.reactionsCount = data.reactionsCount
  }

  get path () {
    return `/posts/${this.id}`
  }

  get textPages () {
    return (this.text || '').split('[newpage]').map(page => page.trim())
  }
}
