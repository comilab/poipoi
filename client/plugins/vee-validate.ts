import Vue from 'vue'
import { ValidationObserver, ValidationProvider, extend, localize } from 'vee-validate'
import * as rules from 'vee-validate/dist/rules'
import ja from 'vee-validate/dist/locale/ja.json'

for (const [rule, validation] of Object.entries(rules)) {
  extend(rule, {
    ...validation as any
  })
}

localize({
  ja
})
localize('ja')

Vue.component('ValidationObserver', ValidationObserver)
Vue.component('ValidationProvider', ValidationProvider)
