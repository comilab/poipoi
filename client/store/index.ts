import { defineComponent, provide, inject } from 'nuxt-composition-api'
import { InjectionKey } from '@vue/composition-api'

import sessionStore from '~/store/session'
import settingStore from '~/store/setting'
import toastStore from '~/store/toast'

type Store = {
  session: ReturnType<typeof sessionStore>
  setting: ReturnType<typeof settingStore>
  toast: ReturnType<typeof toastStore>
}

const makeStore = () => {
  return {
    session: sessionStore(),
    setting: settingStore(),
    toast: toastStore()
  }
}

export const key: InjectionKey<Store> = Symbol('Store')

export const store: Store = makeStore()

export const StoreProvider = defineComponent({
  setup () {
    provide(key, store)
  },
  render (h) {
    return h('div', this.$slots.default)
  }
})

export const useStore = () => inject(key)!

export const resetStore = () => {
  Object.assign(store, makeStore())
}

export const actions = {
  async nuxtClientInit () {
    try {
      await Promise.all([
        store.session.load(),
        store.setting.load()
      ])
    } catch (error) {
      //
    }
  }
}
