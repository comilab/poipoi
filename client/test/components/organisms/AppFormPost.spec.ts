import { fireEvent, render, waitFor } from '@testing-library/vue'
import { mocked } from 'ts-jest/utils'
import { defineComponent, Ref, ref } from '@vue/composition-api'
import { validate } from 'vee-validate'

import AppFormPost from '~/components/organisms/AppFormPost.vue'
import * as store from '~/store'
import usePostsApi from '~/composables/use-posts-api'
import settingFactory from '~/test/factories/setting'
import postFactory from '~/test/factories/post'
import useFontAwesomeIconStub from '~/test/stubs/FontAwesomeIcon'
import useValidationObserverStub from '~/test/stubs/ValidationObserver'
import { fakeEmojis, fakeText, getRules } from '~/test/helper'
import Post from '~/models/Post'

jest.mock('~/composables/use-posts-api')

describe('components/organisms/AppFormPost', () => {
  let options: any
  let usePostsApiMock: any

  beforeEach(() => {
    options = {
      props: {},
      stubs: {
        FontAwesomeIcon: useFontAwesomeIconStub().component,
        VPopover: true,
        AppInputImages: true,
        AppInputEmoji: true
      }
    }

    jest.spyOn(store, 'useStore').mockReturnValue(store.store)
    store.store.setting.setData(settingFactory.build())

    usePostsApiMock = {
      save: jest.fn(),
      post: ref<Post|null>(null)
    }
    usePostsApiMock.save.mockImplementation(() => {
      usePostsApiMock.post.value = postFactory.build()
      return Promise.resolve()
    })
    mocked(usePostsApi).mockImplementation((post) => {
      if (post) {
        usePostsApiMock.post.value = post
      }
      return {
        ...usePostsApiMock
      }
    })
  })

  test('renders correctly', () => {
    store.store.setting.setData(settingFactory.build({
      postDefault: {
        denyRobotScope: [],
        enableReaction: true,
        enableTwitterShare: true,
        allowedEmojis: [],
        deniedEmojis: []
      }
    }))
    const { container } = render(AppFormPost, options)

    expect(container).toMatchSnapshot()
  })

  describe('form', () => {
    describe('inputs', () => {
      describe('タイトル', () => {
        test('32文字まで', async () => {
          const rules = getRules(AppFormPost, 'タイトル', options)!

          const result1 = await validate(fakeText(33), rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.max).toBeTruthy()

          const result2 = await validate(fakeText(32), rules)
          expect(result2.valid).toBeTruthy()
        })
      })

      describe('キャプション', () => {
        test('200文字まで', async () => {
          const rules = getRules(AppFormPost, 'キャプション', options)!

          const result1 = await validate(fakeText(201), rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.max).toBeTruthy()

          const result2 = await validate(fakeText(200), rules)
          expect(result2.valid).toBeTruthy()
        })
      })

      describe('本文', () => {
        test('300000文字まで', async () => {
          const rules = getRules(AppFormPost, '本文', options)!

          const result1 = await validate(fakeText(300001), rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.max).toBeTruthy()

          const result2 = await validate(fakeText(300000), rules)
          expect(result2.valid).toBeTruthy()
        })
      })

      describe('パスワード', () => {
        beforeEach(() => {
          options.props.post = postFactory.build({
            scope: 'password'
          })
        })

        test('scope = passwordなら必須', async () => {
          const { getByTestId } = render(AppFormPost, options)
          expect(getByTestId('input-パスワード')).toBeInTheDocument()

          const rules = getRules(AppFormPost, 'パスワード', options)!

          const result1 = await validate('', rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.required).toBeTruthy()

          const result2 = await validate('password', rules)
          expect(result2.valid).toBeTruthy()
        })

        test('alpha_dashのみ', async () => {
          const rules = getRules(AppFormPost, 'パスワード', options)!

          const result1 = await validate('パスワード', rules)
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.alpha_dash).toBeTruthy()

          const result2 = await validate('password1_-', rules)
          expect(result2.valid).toBeTruthy()
        })
      })

      describe('公開期間', () => {
        test('publishStart < publishEnd', async () => {
          const rules = getRules(AppFormPost, 'publishStart', options)!

          const result1 = await validate('2020-01-01T00:00', rules, {
            values: {
              publishStart: '2020-01-01T00:00',
              publishEnd: '2020-01-01T00:00'
            }
          })
          expect(result1.valid).toBeFalsy()
          expect(result1.failedRules.start_lt_end).toBe('終了日時は開始日時より後にしてください')

          const result2 = await validate('2020-01-01T00:00', rules, {
            values: {
              publishStart: '2020-01-01T00:00',
              publishEnd: '2020-01-01T00:01'
            }
          })
          expect(result2.valid).toBeTruthy()
        })

        test('publishStartかpublishEndが空ならOK', async () => {
          const rules = getRules(AppFormPost, 'publishStart', options)!

          const result1 = await validate('2020-01-01T00:00', rules, {
            values: {
              publishStart: null,
              publishEnd: '2020-01-01T00:00'
            }
          })
          expect(result1.valid).toBeTruthy()

          const result2 = await validate('2020-01-01T00:00', rules, {
            values: {
              publishStart: '2020-01-01T00:00',
              publishEnd: null
            }
          })
          expect(result2.valid).toBeTruthy()
        })
      })

      describe('絵文字', () => {
        describe('デフォルト設定を使用', () => {
          test('初期状態ではtrue', () => {
            const { getByRole } = render(AppFormPost, options)

            expect(getByRole('form')).toHaveFormValues({
              デフォルト設定を使用: true
            })
          })

          test('利用する絵文字か利用しない絵文字がnullでなければfalse', () => {
            options.props.post = postFactory.build({
              allowedEmojis: [],
              deniedEmojis: null
            })
            const { getByRole } = render(AppFormPost, options)
            expect(getByRole('form')).toHaveFormValues({
              デフォルト設定を使用: false
            })
          })

          test('trueならAppInputEmojiが使用不可', () => {
            const inputEmojiProps: any[] = []
            options.stubs.AppInputEmoji = defineComponent({
              props: ['disabled'],
              setup (props) {
                inputEmojiProps.push(props)
              },
              template: '<div><slot name="button" /></div>'
            })
            const { getByRole } = render(AppFormPost, options)

            expect(getByRole('form')).toHaveFormValues({
              デフォルト設定を使用: true
            })
            expect(getByRole('button', { name: '利用する絵文字を選択' })).toHaveAttribute('disabled')
            expect(getByRole('button', { name: '利用しない絵文字を選択' })).toHaveAttribute('disabled')
            expect(inputEmojiProps).toStrictEqual([
              { disabled: true },
              { disabled: true }
            ])
          })

          test('trueなら絵文字一覧にグローバル設定のリストを表示する', () => {
            const inputEmojiProps: any[] = []
            options.stubs.AppInputEmoji = defineComponent({
              props: ['value'],
              setup (props) {
                inputEmojiProps.push(props)
              },
              template: '<div />'
            })
            const { getByRole } = render(AppFormPost, options)

            expect(getByRole('form')).toHaveFormValues({
              デフォルト設定を使用: true
            })
            expect(inputEmojiProps).toStrictEqual([
              { value: store.store.setting.data.value?.postDefault.allowedEmojis },
              { value: store.store.setting.data.value?.postDefault.deniedEmojis }
            ])
          })

          test('falseなら絵文字一覧に投稿設定のリストを表示する', () => {
            const inputEmojiProps: any[] = []
            options.stubs.AppInputEmoji = defineComponent({
              props: ['value'],
              setup (props) {
                inputEmojiProps.push(props)
              },
              template: '<div />'
            })
            options.props.post = postFactory.build({
              allowedEmojis: fakeEmojis(),
              deniedEmojis: fakeEmojis()
            })
            const { getByRole } = render(AppFormPost, options)

            expect(getByRole('form')).toHaveFormValues({
              デフォルト設定を使用: false
            })
            expect(inputEmojiProps).toStrictEqual([
              { value: options.props.post.allowedEmojis },
              { value: options.props.post.deniedEmojis }
            ])
          })
        })
      })
    })

    describe('submit', () => {
      let invalid: Ref<boolean>

      beforeEach(() => {
        const stub = useValidationObserverStub()
        options.stubs.ValidationObserver = stub.component
        invalid = stub.invalid

        options.stubs.AppInputImages = defineComponent({
          setup () {
            return { upload: jest.fn() }
          },
          template: '<div />'
        })
      })

      test('入力エラーがあれば送信不可', () => {
        invalid.value = true
        const { getByRole } = render(AppFormPost, options)

        expect(getByRole('button', { name: '投稿する' })).toHaveAttribute('disabled')
      })

      test('入力エラーがなければ送信可能', () => {
        const { getByRole } = render(AppFormPost, options)

        expect(getByRole('button', { name: '投稿する' })).not.toHaveAttribute('disabled')
      })

      test('送信中は送信不可', async () => {
        const { getByRole } = render(AppFormPost, options)
        await fireEvent.submit(getByRole('form'))

        expect(getByRole('button', { name: '投稿する' })).toHaveAttribute('disabled')
      })

      test('送信中はindeterminate = true', async () => {
        let buttonProps: any
        options.stubs.AppButton = defineComponent({
          props: ['indeterminate'],
          setup (props) {
            buttonProps = props
          },
          template: '<button />'
        })
        const { getByRole } = render(AppFormPost, options)
        expect(buttonProps.indeterminate).toBeFalsy()

        await fireEvent.submit(getByRole('form'))
        expect(buttonProps.indeterminate).toBeTruthy()
      })
    })

    describe('onSubmit', () => {
      let uploadMock: jest.Mock

      beforeEach(() => {
        uploadMock = jest.fn()
        options.stubs.AppInputImages = defineComponent({
          setup () {
            return { upload: uploadMock }
          },
          template: '<div />'
        })
      })

      test('AppInputImagesのuploadを実行', async () => {
        const { getByRole } = render(AppFormPost, options)
        await fireEvent.submit(getByRole('form'))

        expect(uploadMock).toBeCalled()
      })
    })

    describe('onUploaded', () => {
      let uploadMock: jest.Mock
      let inputImagesFireUploadedFunction: Function

      beforeEach(() => {
        uploadMock = jest.fn()
        options.stubs.AppInputImages = defineComponent({
          setup (_props, { emit }) {
            inputImagesFireUploadedFunction = (args: any) => emit('uploaded', args)
            return { upload: uploadMock }
          },
          template: '<div />'
        })
      })

      test('saveを実行', () => {
        render(AppFormPost, options)
        inputImagesFireUploadedFunction()

        expect(usePostsApiMock.save).toBeCalled()
      })

      test('uploadedイベントで渡された引数をinput.imagesにセット', () => {
        const arg = ['foo']
        render(AppFormPost, options)
        inputImagesFireUploadedFunction(arg)

        expect(usePostsApiMock.save.mock.calls[0][0].images).toBe(arg)
      })

      test('useDefaultEmojisがtrueならallowedEmojisとdeniedEmojisをnullにする', async () => {
        options.props.post = postFactory.build({
          allowedEmojis: [],
          deniedEmojis: []
        })
        const { getByLabelText } = render(AppFormPost, options)
        await fireEvent.update(getByLabelText('デフォルト設定を使用'))
        inputImagesFireUploadedFunction()

        expect(usePostsApiMock.save.mock.calls[0][0].allowedEmojis).toBeNull()
        expect(usePostsApiMock.save.mock.calls[0][0].deniedEmojis).toBeNull()
      })

      test('saveに成功したらsubmittedを発火', async () => {
        const { emitted } = render(AppFormPost, options)
        inputImagesFireUploadedFunction()

        await waitFor(() => {
          expect(emitted().submitted).toHaveLength(1)
          expect(emitted().submitted[0][0]).toBeInstanceOf(Post)
        })
      })
    })
  })
})
