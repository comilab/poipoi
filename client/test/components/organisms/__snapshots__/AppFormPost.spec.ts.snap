// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/AppFormPost renders correctly 1`] = `
<div>
  <span>
    <form
      class="max-w-screen-md mx-auto"
      name="post"
    >
      <section
        class="rounded bg-white my-8 p-4"
        data-testid="section"
      >
        <!---->
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-タイトル"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-0"
          >
            タイトル
          </label>
          <input
            class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
            data-testid="input"
            id="input-0"
            name="タイトル"
            type="text"
          />
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <div
              class="text-gray-500 text-right mt-1"
              data-testid="counter"
            >
              0 / 32
            </div>
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-キャプション"
          placeholder="#を付けるとタグになります"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-1"
          >
            キャプション
          </label>
          <textarea
            class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 p-2"
            data-testid="input"
            id="input-1"
            name="キャプション"
            placeholder="#を付けるとタグになります"
            style="height: auto;"
          />
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <div
              class="text-gray-500 text-right mt-1"
              data-testid="counter"
            >
              0 / 200
            </div>
          </div>
        </div>
      </section>
      <section
        class="rounded bg-white my-8 p-4"
        data-testid="section"
      >
        <h3
          class="text-lg"
          data-testid="title"
        >
          画像
        </h3>
        <appinputimages-stub
          images=""
        />
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-null"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-2"
          >
            <span>
              画像の一覧を表示
            </span>
            <vpopover-stub
              class="inline-block"
              placement="top"
              popper-options="[object Object]"
              trigger="click hover"
            >
              <button
                class="text-blue-600 cursor-pointer focus_outline-none ml-2"
                type="button"
              >
                <span />
              </button>
            </vpopover-stub>
          </label>
          <div
            class="flex flex-wrap -mx-4 -my-2"
            data-testid="input"
          >
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value="true"
              />
              <span
                class="whitespace-nowrap"
              >
                表示する
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value="false"
              />
              <span
                class="whitespace-nowrap"
              >
                表示しない
              </span>
            </label>
          </div>
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-1枚目の画像をサムネイルとして表示"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-3"
          >
            <span>
              1枚目の画像をサムネイルとして表示
            </span>
            <vpopover-stub
              class="inline-block"
              placement="top"
              popper-options="[object Object]"
              trigger="click hover"
            >
              <button
                class="text-blue-600 cursor-pointer focus_outline-none ml-2"
                type="button"
              >
                <span />
              </button>
            </vpopover-stub>
          </label>
          <div
            class="flex flex-wrap -mx-4 -my-2"
            data-testid="input"
          >
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                name="1枚目の画像をサムネイルとして表示"
                type="radio"
                value="true"
              />
              <span
                class="whitespace-nowrap"
              >
                表示する
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                name="1枚目の画像をサムネイルとして表示"
                type="radio"
                value="false"
              />
              <span
                class="whitespace-nowrap"
              >
                表示しない
              </span>
            </label>
          </div>
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
      </section>
      <section
        class="rounded bg-white my-8 p-4"
        data-testid="section"
      >
        <h3
          class="text-lg"
          data-testid="title"
        >
          <span>
            本文
          </span>
          <vpopover-stub
            class="inline-block"
            placement="top"
            popper-options="[object Object]"
            trigger="click hover"
          >
            <button
              class="text-blue-600 cursor-pointer focus_outline-none ml-2"
              type="button"
            >
              <span />
            </button>
          </vpopover-stub>
        </h3>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-本文"
        >
          <!---->
          <textarea
            class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 p-2"
            data-testid="input"
            id="input-4"
            name="本文"
            style="height: auto;"
          />
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <div
              class="text-gray-500 text-right mt-1"
              data-testid="counter"
            >
              0 / 300000
            </div>
          </div>
        </div>
      </section>
      <section
        class="rounded bg-white my-8 p-4"
        data-testid="section"
      >
        <h3
          class="text-lg"
          data-testid="title"
        >
          オプション
        </h3>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-公開範囲"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-5"
          >
            公開範囲
          </label>
          <div
            class="flex flex-wrap -mx-4 -my-2"
            data-testid="input"
          >
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                name="公開範囲"
                type="radio"
                value="public"
              />
              <span
                class="whitespace-nowrap"
              >
                全体公開
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                name="公開範囲"
                type="radio"
                value="password"
              />
              <span
                class="whitespace-nowrap"
              >
                パスワード
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                name="公開範囲"
                type="radio"
                value="private"
              />
              <span
                class="whitespace-nowrap"
              >
                非公開
              </span>
            </label>
          </div>
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
        <!---->
        <div
          class="flex flex-col my-4"
        >
          <label
            class="text-sm font-bold py-2"
          >
            <span>
              公開期間
            </span>
            <vpopover-stub
              class="inline-block"
              placement="top"
              popper-options="[object Object]"
              trigger="click hover"
            >
              <button
                class="text-blue-600 cursor-pointer focus_outline-none ml-2"
                type="button"
              >
                <span />
              </button>
            </vpopover-stub>
          </label>
          <div
            class="flex items-start flex-wrap"
          >
            <div
              class="flex flex-col mb-4 md_mb-0 untouched pristine"
              data-testid="input-publishStart"
            >
              <!---->
              <input
                class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
                data-testid="input"
                id="input-6"
                name="publishStart"
                type="datetime-local"
              />
              <div
                class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
              >
                <!---->
                <!---->
              </div>
            </div>
            <div
              class="self-center mx-2"
            >
              〜
            </div>
            <div
              class="flex flex-col untouched pristine"
              data-testid="input-publishEnd"
            >
              <!---->
              <input
                class="border-b focus_border-blue-400 focus_outline-none transition-colors duration-300 bg-white p-2"
                data-testid="input"
                id="input-7"
                name="publishEnd"
                type="datetime-local"
              />
              <div
                class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
              >
                <div
                  data-testid="errors"
                />
                <!---->
              </div>
            </div>
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-null"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-8"
          >
            <span>
              レーティング
            </span>
            <vpopover-stub
              class="inline-block"
              placement="top"
              popper-options="[object Object]"
              trigger="click hover"
            >
              <button
                class="text-blue-600 cursor-pointer focus_outline-none ml-2"
                type="button"
              >
                <span />
              </button>
            </vpopover-stub>
          </label>
          <div
            class="flex flex-wrap -mx-4 -my-2"
            data-testid="input"
          >
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value=""
              />
              <span
                class="whitespace-nowrap"
              >
                なし
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value="nsfw"
              />
              <span
                class="whitespace-nowrap"
              >
                閲覧注意
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value="r18"
              />
              <span
                class="whitespace-nowrap"
              >
                R-18
              </span>
            </label>
          </div>
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-null"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-9"
          >
            <span>
              トップページの先頭に固定
            </span>
            <vpopover-stub
              class="inline-block"
              placement="top"
              popper-options="[object Object]"
              trigger="click hover"
            >
              <button
                class="text-blue-600 cursor-pointer focus_outline-none ml-2"
                type="button"
              >
                <span />
              </button>
            </vpopover-stub>
          </label>
          <div
            class="flex flex-wrap -mx-4 -my-2"
            data-testid="input"
          >
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value="false"
              />
              <span
                class="whitespace-nowrap"
              >
                固定しない
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value="true"
              />
              <span
                class="whitespace-nowrap"
              >
                固定する
              </span>
            </label>
          </div>
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-null"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-10"
          >
            <span>
              検索避け
            </span>
            <vpopover-stub
              class="inline-block"
              placement="top"
              popper-options="[object Object]"
              trigger="click hover"
            >
              <button
                class="text-blue-600 cursor-pointer focus_outline-none ml-2"
                type="button"
              >
                <span />
              </button>
            </vpopover-stub>
          </label>
          <div
            class="flex flex-wrap -mx-4 -my-2"
            data-testid="input"
          >
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value=""
              />
              <span
                class="whitespace-nowrap"
              >
                デフォルト
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value="true"
              />
              <span
                class="whitespace-nowrap"
              >
                する
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value="false"
              />
              <span
                class="whitespace-nowrap"
              >
                しない
              </span>
            </label>
          </div>
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-null"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-11"
          >
            <span>
              リアクション
            </span>
            <vpopover-stub
              class="inline-block"
              placement="top"
              popper-options="[object Object]"
              trigger="click hover"
            >
              <button
                class="text-blue-600 cursor-pointer focus_outline-none ml-2"
                type="button"
              >
                <span />
              </button>
            </vpopover-stub>
          </label>
          <div
            class="flex flex-wrap -mx-4 -my-2"
            data-testid="input"
          >
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value=""
              />
              <span
                class="whitespace-nowrap"
              >
                デフォルト
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value="true"
              />
              <span
                class="whitespace-nowrap"
              >
                受け付ける
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value="false"
              />
              <span
                class="whitespace-nowrap"
              >
                受け付けない
              </span>
            </label>
          </div>
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
        <div
          class="flex flex-col my-4"
        >
          <label
            class="text-sm font-bold py-2"
          >
            <span>
              リアクションに利用する絵文字
            </span>
            <vpopover-stub
              class="inline-block"
              placement="top"
              popper-options="[object Object]"
              trigger="click hover"
            >
              <button
                class="text-blue-600 cursor-pointer focus_outline-none ml-2"
                type="button"
              >
                <span />
              </button>
            </vpopover-stub>
          </label>
          <div
            class="flex flex-col mb-2 untouched pristine"
            data-testid="input-null"
          >
            <!---->
            <label
              class="cursor-pointer"
              data-testid="input"
            >
              <input
                class="mr-2"
                name="デフォルト設定を使用"
                type="checkbox"
                value="true"
              />
              <span>
                デフォルト設定を使用
              </span>
            </label>
            <div
              class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
            >
              <div
                data-testid="errors"
              />
              <!---->
            </div>
          </div>
          <div
            class="md_grid gap-4 grid-cols-2"
          >
            <appinputemoji-stub
              disabled="true"
              value=""
            />
            <appinputemoji-stub
              class="mt-4 md_mt-0"
              disabled="true"
              value=""
            />
          </div>
        </div>
        <div
          class="flex flex-col my-4 untouched pristine"
          data-testid="input-null"
        >
          <label
            class="text-sm font-bold py-2"
            data-testid="label"
            for="input-13"
          >
            <span>
              Twitterシェア
            </span>
            <vpopover-stub
              class="inline-block"
              placement="top"
              popper-options="[object Object]"
              trigger="click hover"
            >
              <button
                class="text-blue-600 cursor-pointer focus_outline-none ml-2"
                type="button"
              >
                <span />
              </button>
            </vpopover-stub>
          </label>
          <div
            class="flex flex-wrap -mx-4 -my-2"
            data-testid="input"
          >
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value=""
              />
              <span
                class="whitespace-nowrap"
              >
                デフォルト
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value="true"
              />
              <span
                class="whitespace-nowrap"
              >
                有効
              </span>
            </label>
            <label
              class="cursor-pointer mx-4 my-2"
            >
              <input
                class="mr-2"
                type="radio"
                value="false"
              />
              <span
                class="whitespace-nowrap"
              >
                無効
              </span>
            </label>
          </div>
          <div
            class="flex flex-col md_flex-row flex-col-reverse justify-between text-sm"
          >
            <div
              data-testid="errors"
            />
            <!---->
          </div>
        </div>
      </section>
      <div
        class="text-center"
      >
        <button
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-blue-600 text-white hover_bg-blue-700 px-3 py-2 rounded"
          data-testid="button"
          type="submit"
        >
          <div>
            <span
              class="fa-fw mr-2"
              data-testid="icon"
            />
          </div>
          <div
            class=""
            data-testid="text"
          >
            投稿する
          </div>
        </button>
      </div>
    </form>
  </span>
</div>
`;
